### ISSUE TITLE
Sudden rejections with "Build config did not create any jobs"

### ISSUE THREAD

**Issue by achimnol** [travis-ci/travis-ci#8536]
Please check https://travis-ci.org/achimnol/aiotools/requests here.
It's failing after updating `.travis.yml` to use explicit stages. Other my projects that just got the same update are working well with it! [The new config is here.](https://github.com/achimnol/aiotools/blob/78ddd74a9b7e85bfa7e6ca54fbc9dbbbe649052f/.travis.yml)
As it even doesn't create the job, I cannot restart or cancel anything.

**Comment by BanzaiMan** [comment]
Here's what is happening.

With https://github.com/achimnol/aiotools/blob/287a12ea5b8f7fd68de906e7c0e9fc4a6d0099e3/.travis.yml, you define two stages, `test` and `deploy`, but the sole job in `test` is implicit. Since you add an extra job with `jobs.include`, this implicit job is removed.

Recent build requests for https://github.com/achimnol/aiotools are not tagged, the only one that is defined is also removed, thus resulting in an empty build.

To ensure you get at least one job, then, you should explicitly define the `test` job inside. See https://github.com/achimnol/aiotools/pull/10.

**Comment by achimnol** [comment]
Thanks!
Hm.. I just wonder why [this config](https://github.com/aio-libs/aiodocker/blob/f55f3edb6d7c46b2920185e0866d4d1db88acb8e/.travis.yml) and [this config](https://github.com/lablup/backend.ai-client-py/blob/df33c9d2ebbae0a646becdd9340429616ba0d742/.travis.yml) works with implicit test stage.

**Comment by BanzaiMan** [comment]
Because those define at least two jobs in the top-level matrix.

**Comment by achimnol** [comment]
Okay, I got it.
I'd like to request some documentation improvements on how implicit/explicit stages work. I could not spot any mention about "a single job in the implicit test stage is removed" at [the build stage doc](https://docs.travis-ci.com/user/build-stages/).
Also, it mentions a little about inheritance:
<img width="854" alt="image" src="https://user-images.githubusercontent.com/555156/31308677-ad925e94-abb5-11e7-85c1-22dda55eb6bc.png">
but I still don't get how it works.
For instance, consider the following two cases:
```
install: ...
script: ...
jobs:
  include:
    - stage: deploy
```
vs.
```
jobs:
  include:
    - stage: test
      install: ...
      script: ...
    - stage: deploy
      deploy: ...
```
I could understand the deploy stage in the first case inherits install/script fields.
But I'm not sure how it works in the second case. Does the deploy stage in the second case also inherits the test stage?

**Comment by BanzaiMan** [comment]
The reason that the implicit job is removed is explained in this issue: https://github.com/travis-ci/travis-ci/issues/4681 (in particular, https://github.com/travis-ci/travis-ci/issues/4681#issuecomment-306177018). I agree that this is more clearly explained in the document.

The "array" in matrix dimensions (as used in the document) means the values defined for the matrix expansion for each language (e.g., `rvm`, `jdk`, etc. for `language: ruby`; `node_js` for `language: node_js`; `env` for all languages).

To explain with a concrete example, I point to your [original configuration](https://github.com/achimnol/aiotools/blob/287a12ea5b8f7fd68de906e7c0e9fc4a6d0099e3/.travis.yml). It has [an array for `python` (`python: ["3.6"]`)](https://github.com/achimnol/aiotools/blob/287a12ea5b8f7fd68de906e7c0e9fc4a6d0099e3/.travis.yml#L11-L12). The initial value is inherited by the jobs included in `jobs.include`. If your `python` key had multiple values, say `python: ["3.5", "3.6"]` at the top level, no jobs will be removed and all jobs in `jobs.include` will define `python: "3.5"`.

**Comment by achimnol** [comment]
Thanks for more detailed explanation.
I hope to see improvements in documentation so that newbies like me will have less time to struggle with repeating commit-and-test cycles.

**Comment by Time0o** [comment]
Sorry for bringing this up again but I'm getting the rejection message for the following `.travis.yml` which is essentially identical to the official `Running Python tests on multiple Operating Systems` example. Is the documentation out of date or is something else going on here? From the discussion above I don't really get how I would have to change this.

```
language: python

matrix:
    include:
        - name "Xenial"
          dist xenial
        - name "maxOS"
          os osx
          os_image xcode10.2
          language shell

install: ./.travis_install.sh

script: conda activate colorization-pytorch || ./run_tests.py
```

**Comment by vgvassilev** [comment]
In some cases the fix of travis-ci/issues#4681 leads to the issue travis-ci/issues#8536. Removing the excludes clause makes it work again...
